Implementation of BIP 101 : maximum block size increase

gavinandresen opened the pull request:
Transaction volume is approaching the one-megabyte blocksize limit. We will never reach that limit, because several things will happen as we get close:
1. Transaction confirmation times for transactions with a given fee will rise; very-low-fee transactions will fail to get confirmed at all.
2. Average transaction fee paid will rise
3. People or applications unwilling or unable to pay the rising fees will stop submitting transactions
4. People and businesses will shelve plans to use Bitcoin, stunting growth and adoption

That will be a major economic shift from the status-quo of low transaction fees. See the preliminary BIP100 proposal: http://gtf.org/garzik/bitcoin/BIP100-blocksizechangeproposal.pdf for more discussion.

There is rough (but not 100%) consensus that at some point fees will rise "high enough" that the 1MB blocksize limit will have to be raised, but no consensus on what that point should be and no consensus on any process to determine what that point should be.

[BIP 101](https://github.com/bitcoin/bips/pull/163) and the changes in this pull request are intended to keep the economic status quo (and keep Bitcoin growing) while acknowledging the engineering reality that networks do not have infinite bandwidth.

I addressed technical objections in a series of blog posts at http://gavinandresen.ninja/time-to-roll-out-bigger-blocks .

Any consensus rule change should have thorough code review and time for testing, so I am submitting this code now so that it has thorough code review and testing before transaction volume gets closer to the one megabyte limit.

Higher-level discussions on the merits of BIP 101 versus other proposals belongs on the bitcoin-dev mailing list (which just moved: https://lists.linuxfoundation.org/mailman/listinfo/bitcoin-dev )-- please limit the conversation here to discussion of the code changes, and rest assured these changes will not be pulled into Bitcoin Core unless there is consensus around BIP 101.

gavinandresen committed:
Refactor, new CNode::FinalizeHeader method

I need this to write some unit tests for the
CNode::ReceiveMsgBytes() function.

gavinandresen committed:
Unit test for CNode::ReceiveMsgBytes

gavinandresen committed:
Allow per-message sanity checking when reading from wire

gavinandresen committed:
Implement hard fork to allow bigger blocks

Unit test and code for a bigger-block hard fork.
Parameters are:
  8MB cap
  ... doubling every two years (so 16MB in 2018)
  ... for twenty years
  ... earliest possible chain fork: 11 Jan 2016
  ... after miner supermajority (code in the next patch)
  ... and grace period once miner supermajority achieved (code in next patch)

gavinandresen committed:
Implement miner vote and grace period for block size fork

These changes implement and test miner rollout of a bigger
block size hard fork.

qa/rpc-tests/bigblocks.py mines a 50% upgraded -regtest chain,
generates 4MB of transactions, and then tests the upgrade and
mining code to make sure larger blocks are only created when
the fork conditions are met.

The activation state of the fork is stored in the block tree database;
it is written when the threshold is met (and unwritten if the
threshold block is re-orged out of the best chain), and read at startup.

gavinandresen committed:
Define MIN_TRANSACTION_SIZE=60 constant (avoid repeating myself)

gavinandresen commented:
The per-message size sanity checking commit is also pull request #6261 

The mocktime fixes are pull request #6337 

The remove coinbase payment key from keypool commit will be obsolete if #5994 is merged

juiced commented:
What about lowering the blocksize to 512KB first and see if everything happens what we think will happen? /me runs away

btcdrak commented on src/chainparams.cpp:
```diff
@@ -53,6 +53,20 @@ class CMainParams : public CChainParams {
+        nSizeDoubleEpoch = 60*60*24*365*2; // two years
+        nMaxSizeBase = 8*1000*1000; // 8MB
+        nMaxSizeDoublings = 10;
+        nActivateSizeForkMajority = 750; // 75% of hashpower to activate fork
```
This is too little. Softforks reach the point of no return at 95%

btcdrak commented on src/chainparams.cpp:
```diff
@@ -150,6 +164,17 @@ class CTestNetParams : public CMainParams {
+        nSizeDoubleEpoch = 60*60*24*365*2; // two years
+        nMaxSizeBase = 8*1000*1000; // 8MB
+        nMaxSizeDoublings = 10;
+        nActivateSizeForkMajority = 75; // 75 of 100 to activate fork
```
This is too little. Softforks reach the point of no return at 95%

rebroad replied:
I agree. (Although, with my merge-mine proposal below, this becomes less relevant. 50% might be sufficient to start the merge-mined fork).

rebroad commented:
Hi @gavinandresen, I've emailed you about this, but I might as well mention it here. There a lot of people who would like to see bitcoin with its 1MB block limit continue to exist. I am in favour of a hard fork to allow bitcoin to scale, but only if it doesn't destroy the 1MB-limit bitcoin in the process. I would like to see how both evolve rather than having to choose between them.

Therefore, since you are already proposing a hard fork, why not make the hard fork merge mine with the existing bitcoin - this way miners can continue to mine both coins. Our coins in each coin will be safe, and whatever happens to the market cap - i.e. whichever prospers, everyone will have hedged their bets since their coins will exist in both forks (at the point of the fork). I think it is far more likely that the market cap combined for the two coins will suffer far less than the market cap of just one coin continuing - since not everyone will have confidence in the coin chosen, whereas far more people are likely to have confidence in one OR the other.

Publicity around this could suggest who goes with which coin. Exchanges of course would want to continue allowing trading for both, but most merchants I'd expect would want to default to the scalable bitcoin. The current/original 1MB bitcoin will in time become associated to be more gold-like (for long-term investments), whereas the scaled bitcoin will in time become more associated with money and day-to-day transactions.

I have yet to hear any logical arguments against this proposal - other than "publicity nightmare", which I think can equally (or more) apply to a solution which disallows either the 1MB coin to live or disallows the scaled coin to be born.

btcdrak commented on src/chainparams.cpp:
```diff
@@ -150,6 +164,17 @@ class CTestNetParams : public CMainParams {
+        // 1MB max blocks before 1 Aug 2015
+        // Then, if miner consensus: 8MB max, doubling every two years
+        nMaxSizePreFork = 1000*1000; // 1MB max pre-fork
+        nEarliestSizeForkTime = 1438387200; // 1 Aug 2015 00:00:00 UTC
```
There should be a nLatestSizeForkTime which is the latest the fork can be activated, and if not, the fork would fail.

rnicoll commented:
@rebroad From a technical point of view I'd need to check, but I think having a coin merge-mine with its unforked self is technically viable. As the old chain won't have a chain ID, and the new chain will, there shouldn't be any issues there.

Non-technical issues, however - because intentionally splitting the value of Bitcoin between two chains is unlikely to be popular (now you need two chains to keep your entire value pre-fork), because service providers either have to pick one or adopt a second chain (which many of them are not set up to handle), because it's a branding nightmare of trying to explain which Bitcoin you accept, and most importantly for those of us who believe a 1MB chain is doomed to failure, it would essentially be reckless to be an enabler for such an approach.

There is part of me that wants to make consensus parameters runtime configurable and stand well back, but realistically doing so would likely be extremely damaging, and for similar reasons I don't think encouraging people to split is a good idea if we can find an approach most will accept.

mjamin commented:
> Higher-level discussions on the merits of BIP 101 versus other proposals belongs on the bitcoin-dev mailing list

petertodd commented:
Note to readers: in its current form there is a near zero chance of this getting merged due to a number of BIP-level issues in addition to debate about the patch itself. For instance, Gavin has never given any details about testing; at minimum we'd need a BIP16 style quality assurance document. We also frown on writing software with building expiration dates, let alone expiration dates that trigger non-deterministically. (Note how my recently merged CLTV considered the year 2038 problem to avoid needing a hard fork at that date)

Until these issues are addressed I an many other contributors will be muting this thread and ignoring comments until the BIP itself is fixed. Much of the discussion we see in conversations around this subject is highly repetitive and a big timesink; don't interpret silence as agreement.

sipa commented:
> There is rough (but not 100%) consensus that at some point fees will rise "high enough" that the 1MB blocksize limit will have to be raised, but no consensus on what that point should be and no consensus on any process to determine what that point should be.

See the mailinglist for why I totally disagree with the fact that fear of higher fees are a reason to increase the block size. If fees rise high enough, and people pay them, that sounds like a fantastic problem to have...

LeMiner commented:
@sipa , without the approapriate sized userbase there won't be anyone interested in paying fees on a network that can only support 200.000 people, nor will there be any economical incentive to continue payments on such a network. 

Growth attracts users and users create transactions, it's the developers job to make sure that the network can grow and support as many users as possible, not to make decisions on where to limit the network or on how many transactions it can support.

laanwj commented:
I'm happy to see the code for review openly here, but definitely do **not** take the block size discussion here.

theuni commented on src/chainparams.h:
```diff
@@ -71,14 +72,62 @@ class CChainParams
+    /** Maximum block size of a block with timestamp nBlockTimestamp */
+    int ActivateSizeForkMajority() const { return nActivateSizeForkMajority; }
+    uint64_t SizeForkGracePeriod() const { return nSizeForkGracePeriod; }
+    uint64_t MaxBlockSize(uint64_t nBlockTimestamp) const {
```
Could you please pass nSizeForkActivationTime in here, rather than changing the const assumption of Params()? main can be responsible for keeping tabs of the activation time.

It will make life much easier for libbitcoinconsensus. I should think that it would make it easier for testing as well, since you could more easily fake a time for a single test.

theuni commented on src/main.cpp:
```diff
@@ -1870,7 +1880,7 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
         nInputs += tx.vin.size();
         nSigOps += GetLegacySigOpCount(tx);
-        if (nSigOps > MAX_BLOCK_SIGOPS)
+        if (nSigOps > Params().MaxBlockSigops(block.GetBlockTime()))
```
Please use chainparams here rather than Params().

theuni commented on src/main.cpp:
```diff
@@ -1886,7 +1896,7 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
                 // an incredibly-expensive-to-validate block.
                 nSigOps += GetP2SHSigOpCount(tx, view);
-                if (nSigOps > MAX_BLOCK_SIGOPS)
+                if (nSigOps > Params().MaxBlockSigops(block.GetBlockTime()))
```
Here too, and wherever else it applies. I'll avoid pointing out any others.

theuni commented:
Only had a minute to look over it, but I added a few comments about code for the sake of reviewing code. Policies and numbers can be discussed elsewhere, it's reasonable to critique the code independent of those discussions.

I'll give it a thorough review over the weekend.

gavinandresen commented:
I'm going to delete a few comments from @btcdrak and @rebroad that don't follow the "comment on the code-- high-level discussion goes to bitcoin-dev mailing list"

@theuni : thanks for the review-- this was forward-ported from 0.10 to 0.11 to git HEAD. It would be helpful if there was a design document for libbitcoinconsensus that I could code against...

theuni commented:
@gavinandresen Point taken. I'm not sure if we're quite to that point yet, though. We're still in the "move things around so we can do something useful" part and not quite to the "what useful things to do and how" part.

In particular for this PR, consensus functions need to work statelessly, so something like CheckTransaction() should have deterministic results depending only on its parameters and chainparam constants. That is a work-in-progress atm, see https://github.com/bitcoin/bitcoin/pull/6051 as an example of the current momentum. FWIW, that's why we've started passing chainparams and consensusparams around everywhere.

For the scope of this PR i don't think it's necessary to worry about that too much, but it'd be nice if it didn't make chainparams itself stateful.

That said, these new vars/functions probably belong in consensus/params.h rather than chainparams anyway; that's the new home for consensus-critical stuff.

I'd prefer to see something like:
```patch
diff --git a/src/consensus/params.h b/src/consensus/params.h
index c480a1c..b4b97f6 100644
--- a/src/consensus/params.h
+++ b/src/consensus/params.h
@@ -24,7 +24,22 @@ struct Params {
     bool fPowAllowMinDifficultyBlocks;
     int64_t nPowTargetSpacing;
     int64_t nPowTargetTimespan;
+
+    /** Maximum block size parameters */
+    uint32_t nMaxSizePreFork;
+    uint64_t nEarliestSizeForkTime;
+    uint32_t nSizeDoubleEpoch;
+    uint64_t nMaxSizeBase;
+    uint8_t nMaxSizeDoublings;
+    int nActivateSizeForkMajority;
+    uint64_t nSizeForkGracePeriod;
+
     int64_t DifficultyAdjustmentInterval() const { return nPowTargetTimespan / nPowTargetSpacing; }
+    uint64_t MaxTransactionSize(uint64_t nBlockTimestamp, uint64_t nForkTime) const { ... };
+    uint64_t MaxBlockSize(uint64_t nBlockTimestamp, uint64_t nForkTime) const { ... };
+    uint64_t MaxBlockSigops(uint64_t nBlockTimestamp, uint64_t nForkTime) const { return MaxBlockSize(nBlockTimestamp, nForkTime)/50; }
+
+    }
 };
 } // namespace Consensus

```

sandakersmann commented:
concept ACK

gavinandresen committed:
Move size fork params from chainparams.h to consensus/params.h

gavinandresen commented:
Reworked to fix @theuni code review comments:

+ Made chainparams constant: main.cpp holds a sizeForkTime global (that behaves like a std::atomic)
+ Moved block size fork params/methods to consensus/params.h

Regarding your comment on the change to over-maximum-block sizes:

That is a behavior change, but only if you send blocks between 1 and 2 megabytes before the fork switchover.  Before: Send a block message > 2 MB: you get immediately disconnected (but not banned). Between 1 and 2MB: CheckBlock() would call Misbehaving() and you might get disconnected/banned (if not whitelisted).

With this pull request: any oversize block will get you immediately disconnected.

I started to make the behavior more consistent... but it's complicated because Misbehaving() requires the cs_main lock be held, but the SanityCheckMessage() callback is deep inside the network-servicing code with cs_vRecvMsg lock held.  Untangling that is, I think, a project for another day.

gavinandresen commented:
@theuni : new Travis error:
```{standard input}: Fatal error: can't close test/test_test_bitcoin-sanity_tests.o: No space left on device```

Is there anything I can do to fix that?

theuni commented:
@gavinandresen given that it only happened on the one builder and i've never seen it before, I'd chalk that up to a freak occurrence unless it becomes a trend. Fire it off again?

gavinandresen commented:
It was, indeed, a spurious Travis error.

A note on the Java comparison tool: I believe it will continue to work with this pull request "forever" because it produces down-version blocks for testing; it will never trigger the supermajority-of-miners condition to make bigger blocks legal.

dgenr8 commented:
Been running on mainnet for 1 day.  Got a lot of "Oversized addr" disconnects (65% of all connections, in bursts).  Don't know what was being received though.  Also got one "Oversized headers" disconnect.

gavinandresen committed:
Refactor: protect mapNodeState with its own lock

Encapsulate mapNodeState in a smart-pointer class with its own lock.

Why? So Misbehaving() can be called from the sanity-check-a-message code without holding cs_main.

And to get better exception safety (the smart-pointer approach gives RAII semantics).

And because protecting fewer things with cs_main is a good idea.

Tested by compiling with -DDEBUG_LOCKORDER, running all of the qa/rpc-tests, and running a node on the main network overnight.

gavinandresen commented:
Closing for now; too painful to try to keep up with the movement of code from main.h to policy/policy.h.

jeffreyrufino commented:
Do you guys think this has affected the Bitcoin Protocol?

ralyodio reviewed (commented):
LGTM
